File: problem_045.c
// -*- compile-command: "gcc -o problem_045 problem_045.c ceuler.c -Wall -lm" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>
#include <stdlib.h>
#include "ceuler.h"

int main()
{
    printf("Project Euler - Problem 45:\n"
           "Find the next triangle number after 40755 that is also pentagonal and hexagonal.\n\n");

    long long number  = 40755; // Starting Triangle number
    int triHelp = 286;

    while ( true )
    {
        // get next triangle number
        number += triHelp;
        triHelp++;

        if ( isPentagonal(number) != true ) continue;
        if ( isHexagonal(number)  != true ) continue;
        break;
    }
    printf("Next trigonal, pentagonal, and hexagonal number: %lld\n", number);
    return 0;
}

File: problem_045.rs
// -*- compile-command: "rustc -o problem_045_rs problem_045.rs" -*-
// Copyright (c) 2016 Michael Caldwell

fn main() {
    println!("Project Euler - Problem 45:");
    println!("Find the next triangle number after 40755 that is also pentagonal and hexagonal.\n");

    let mut tri: [u64; 2]  = [285, 40755];
    let mut pent: [u64; 2] = [165, 0];
    let mut hex: [u64; 2]  = [143, 0];

    while tri[1] != hex[1] || tri[1] != pent[1] {
        // Get next Triangle number
        tri[0] += 1;
        tri[1] = tri[0] * (tri[0] + 1) / 2;

        // Get next Pentagonal number once Triangle number is larger
        while pent[1] < tri[1] {
            pent[0] += 1;
            pent[1] = pent[0] * (3 * pent[0] - 1) / 2;
        }

        // Get next Hexagonal number once Triangle number is larger
        while hex[1] < tri[1] {
            hex[0] += 1;
            hex[1] = hex[0] * (2 * hex[0] - 1);
        }
    }

    println!("{}", tri[1]);
}
